pkg/admission/server.go

```diff
@@ -0,0 +1,161 @@
+package admission
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	admission "k8s.io/api/admission/v1"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/klog/v2"
+	"net/http"
+	"path/filepath"
+	"reflect"
+	v1alpha1 "sigs.k8s.io/service-apis/apis/v1alpha1"
+)
+
+const (
+	tlsDir      = `/run/secrets/tls`
+	tlsCertFile = `tls.crt`
+	tlsKeyFile  = `tls.key`
+)
+
+var (
+	scheme = runtime.NewScheme()
+	codecs = serializer.NewCodecFactory(scheme)
+)
+
+func log500(w http.ResponseWriter, err error) {
+	klog.Errorf("failed to read request from client: %v\n", err)
+	http.Error(w, err.Error(), http.StatusInternalServerError)
+	return
+}
+
+// ServeHTTP parses AdmissionReview requests and responds back
+// with the validation result of the entity.
+func ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if r.Body == nil {
+		klog.Errorf("received request with empty body")
+		http.Error(w, "admission review object is missing",
+			http.StatusBadRequest)
+		return
+	}
+	data, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		log500(w,err)
+	}
+	review := admission.AdmissionReview{}
+	if err := json.Unmarshal(data, &review); err != nil {
+		klog.Errorf("failed to parse AdmissionReview object: %v\n", err)
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+	response, err := handleValidation(*review.Request)
+	if err != nil {
+		log500(w,err)
+	}
+	review.Response = response
+	data, err = json.Marshal(review)
+	if err != nil {
+		log500(w,err)
+	}
+	_, err = w.Write(data)
+	if err != nil {
+		log500(w,err)
+	}
+}
+
+var (
+	httpRoute = meta.GroupVersionResource{
+		Group:    v1alpha1.SchemeGroupVersion.Group,
+		Version:  v1alpha1.SchemeGroupVersion.Version,
+		Resource: "httproutes",
+	}
+	gateway = meta.GroupVersionResource{
+		Group:    v1alpha1.SchemeGroupVersion.Group,
+		Version:  v1alpha1.SchemeGroupVersion.Version,
+		Resource: "gateway",
+	}
+)
+
+func routesEqual(new v1alpha1.HTTPRoute, old v1alpha1.HTTPRoute) bool {
+	if reflect.DeepEqual(new, old) {
+		return true
+	}
+	return false
+}
+
+func handleValidation(request admission.AdmissionRequest) (
+	*admission.AdmissionResponse, error) {
+	var response admission.AdmissionResponse
+	var ok bool
+	var message string
+	var err error
+
+	if request.Resource == httpRoute {
```

Comment:
This might be a taste thing but we should handle all resources. Using a `switch` statement might help here.

Reply:
Should we default to skipping validation if it's not specifically the resources that we are targeting ?

Reply:
Yes, I think so. The `default` case of the switch statement can be used for that. A debug log in there would be helpful.

Reply:
My only thought on this is that the `if` statement means that we do not have to have fallthrough for each case so that we handle multiple validations in giant yaml files.

Reply:
> My only thought on this is that the if statement means that we do not have to have fallthrough for each case so that we handle multiple validations in giant yaml files.

I'm not sure I understand. Are you referring to handling cases by resources or something more fine-grained?

Reply:
I think I confused myself, I assume a giant file with mulitple httpRoutes in it would not be validated completely but I don't think that's the case.